### PR TITLE
clients: feature gate `tls`

### PR DESCRIPTION
**Pull request by niklasad1**
Closing #241

### CODE REVIEW

File: tests/tests/integration_tests.rs [hunk 1]
```diff
@@ -245,7 +245,7 @@ async fn https_works() {
 #[tokio::test]
 #[ignore]
 async fn wss_works() {
-	let client = WsClientBuilder::default().build("wss://kusama-rpc.polkadot.io").await.unwrap();
+	let client = WsClientBuilder::default().build("wss://kusama-rpc.polkadot.io:443").await.unwrap();
```
Comment (niklasad1) [inline tests/tests/integration_tests.rs:248]:
```suggestion
	let client = WsClientBuilder::default().build("wss://kusama-rpc.polkadot.io").await.unwrap();
```

File: ws-client/src/stream.rs [hunk 2]
```diff
@@ -32,30 +32,29 @@ use futures::{
...
 use std::{io::Error as IoError, pin::Pin, task::Context, task::Poll};
+use tokio::net::TcpStream;
 use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
 
 /// Stream to represent either a unencrypted or encrypted socket stream.
 #[pin_project(project = EitherStreamProj)]
-#[derive(Debug, Copy, Clone)]
-pub enum EitherStream<S, T> {
+#[derive(Debug)]
+pub enum EitherStream {
```
Comment (niklasad1) [inline ws-client/src/stream.rs:41]: I did this but I realized I can just used placeholder type such as `()` to keep the generics here not sure that we need the generics anyway.

Unless we plan to support any transport using `tokio rustls` :)

File: ws-client/src/transport.rs [hunk 3]
```diff
@@ -169,38 +166,46 @@ impl Receiver {
...
+		};
+
 		for _ in 0..self.max_redirections {
 			tracing::debug!("Connecting to target: {:?}", target);
 
 			// The sockaddrs might get reused if the server replies with a relative URI.
 			let sockaddrs = std::mem::take(&mut target.sockaddrs);
 			for sockaddr in &sockaddrs {
-				let tcp_stream = match connect(*sockaddr, self.timeout, &target.host, &tls_connector).await {
+				#[cfg(feature = "tls")]
```
Comment (niklasad1) [inline ws-client/src/transport.rs:189]: I had to do this to avoid feature gate the entire method.

File: http-client/Cargo.toml [hunk 4]
```diff
@@ -26,3 +26,7 @@ url = "2.2"
 [dev-dependencies]
 jsonrpsee-test-utils = { path = "../test-utils" }
 tokio = { version = "1.8", features = ["net", "rt-multi-thread", "macros"] }
+
+[features]
+default = ["tls"]
+tls = ["hyper-rustls/webpki-tokio"]
```
Comment (maciejhirsz) [inline http-client/Cargo.toml:32]: TIL

**Review by maciejhirsz — APPROVED** [review]
Looks good, do we have CI that tests with different features on and off?

**Comment by niklasad1** [comment]
> Looks good, do we have CI that tests with different features on and off?

yeah, we have [this](https://github.com/paritytech/jsonrpsee/blob/master/.github/workflows/ci.yml#L70) which should compile each crate with `default-features = false` and test each feature "standalone". However we just run the tests on the `default features` because it was so slow but I can double check the tests locally to be on the safe-side

File: http-client/Cargo.toml [hunk 5]
```diff
@@ -13,16 +13,19 @@ documentation = "https://docs.rs/jsonrpsee-http-client"
...
-hyper-rustls = { version = "0.23", features = ["webpki-tokio"] }
+hyper-rustls = { version = "0.23", optional = true }
 jsonrpsee-types = { path = "../types", version = "0.6.0" }
 jsonrpsee-utils = { path = "../utils", version = "0.6.0", features = ["client", "http-helpers"] }
 serde = { version = "1.0", default-features = false, features = ["derive"] }
 serde_json = "1.0"
 thiserror = "1.0"
 tokio = { version = "1.8", features = ["time"] }
 tracing = "0.1"
-url = "2.2"
```
Comment (niklasad1) [inline http-client/Cargo.toml:24]: This will close https://github.com/paritytech/jsonrpsee/issues/554

**Review by dvdplm — APPROVED** [review]
lgtm

**Pull request merged**